File: backend/Ludikore.Revoicer.Services/Cloud/CloudProviderFactory.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Ludikore.Revoicer.Services.AWS;
using Ludikore.Revoicer.Services.Azure;
using Microsoft.Extensions.Configuration;

namespace Ludikore.Revoicer.Services.Cloud
{
    /// <summary>
    /// This class returns preconfigured instances of the cloud services for the cloud provider specified on its initialization
    /// </summary>
    public class CloudProviderFactory
    {


        /// <summary>
        /// Gets the cloud provider being used.
        /// </summary>
        /// <value>The cloud provider.</value>
        public CloudProvider Provider { get; }

        public CloudSettings CloudSettings { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="CloudProviderFactory" /> class.
        /// </summary>
        /// <param name="provider">The provider.</param>
        /// <param name="configuration"></param>
        /// <param name="cloudSettings">The settings.</param>
        public CloudProviderFactory(CloudProvider provider, IConfiguration configuration)
        {
            Provider = provider;
            CloudSettings = new CloudSettings(configuration);
        }

        /// <summary>
        /// Creates or retrieves an instance of the service that provides access to message queues
        /// </summary>
        /// <returns>CloudQueueService.</returns>
        /// <exception cref="Ludikore.Revoicer.Services.Cloud.RevoicerConfigurationException">There is no entry for this cloud provider: {Provider}</exception>
        public CloudQueueService GetQueueService()
        {
            switch (Provider)
            {
                case CloudProvider.AWS:
                    return new SqsService(CloudSettings.AwsEndpoint);
                case CloudProvider.Azure:
                    var queueConnectionString = $"Endpoint={CloudSettings.AzureServiceBusEndpoint};" +
                                                $"SharedAccessKeyName={CloudSettings.AzureServiceBusAccessKeyName};" +
                                                $"SharedAccessKey={CloudSettings.AzureServiceBusAccessKeyValue};" +
                                                $"AccountName={CloudSettings.AzureAccountName}";

                    return new ServiceBusService(queueConnectionString);
                default:
                    throw new RevoicerConfigurationException($"There is no entry for this cloud provider: {Provider}");
            }
        }

        /// <summary>
        /// Creates or retrieves an instance of the service that provides access to the cloud storage
        /// </summary>
        /// <returns>CloudStorageService.</returns>
        /// <exception cref="Ludikore.Revoicer.Services.Cloud.RevoicerConfigurationException">There is no entry for this cloud provider: {Provider}</exception>
        public CloudStorageService GetStorageService()
        {
            switch (Provider)
            {
                case CloudProvider.AWS:
                    return new S3Service("http://localstack:4566");
                case CloudProvider.Azure:

                    var storageConnectionString = "DefaultEndpointsProtocol=https;" +
                                                  $"AccountName={CloudSettings.AzureAccountName};" +
                                                  $"AccountKey={CloudSettings.AzureStorageAccessKey};" +
                                                  "EndpointSuffix=core.windows.net";

                    return new BlobStorageService(
                        connectionString: storageConnectionString,
                        sasAccountKey: CloudSettings.AzureStorageAccessKey);
                default:
                    throw new RevoicerConfigurationException($"There is no entry for this cloud provider: {Provider}");
            }
        }
    }
}

File: workers/common/cloud/CloudProviderFactory.py
import os

from workers.common.azure.BlobStorageService import BlobStorageService
from workers.common.azure.ServiceBusService import ServiceBusService
from workers.common.cloud.CloudProvider import CloudProvider
from workers.common.cloud.CloudQueueService import CloudQueueService
from workers.common.cloud.CloudStorageService import CloudStorageService


AZURE_SB_ENDPOINT = os.environ["REVOICER_AZURE_SB_ENDPOINT"]
AZURE_SB_ACCESS_KEY_NAME = os.environ["REVOICER_AZURE_SB_ACCESS_KEY_NAME"]
AZURE_SB_ACCESS_KEY_VALUE = os.environ["REVOICER_AZURE_SB_ACCESS_KEY_VALUE"]

AZURE_ACCOUNT_NAME = os.environ["REVOICER_AZURE_ACCOUNT_NAME"]
AZURE_STORAGE_ACCESS_KEY = os.environ["REVOICER_AZURE_STORAGE_ACCESS_KEY"]
AZURE_STORAGE_CONTAINER_NAME = os.environ["REVOICER_AZURE_STORAGE_CONTAINER_NAME"]

# Settings
QUEUE_CONNECTION_STRING = f"Endpoint={AZURE_SB_ENDPOINT};" \
                          f"SharedAccessKeyName={AZURE_SB_ACCESS_KEY_NAME};" \
                          f"SharedAccessKey={AZURE_SB_ACCESS_KEY_VALUE};" \
                          f"AccountName={AZURE_ACCOUNT_NAME};"

STORAGE_CONNECTION_STRING = f"DefaultEndpointsProtocol=https;" \
                            f"AccountName={AZURE_ACCOUNT_NAME};" \
                            f"AccountKey={AZURE_STORAGE_ACCESS_KEY};" \
                            f"EndpointSuffix=core.windows.net;"
class CloudProviderFactory:
    def __init__(self, cloud_provider: CloudProvider):
        self.cloud_provider = cloud_provider

    def get_queue_service(self) -> CloudQueueService:
        match self.cloud_provider:
            case CloudProvider.Azure:
                return ServiceBusService(QUEUE_CONNECTION_STRING)
            case _:
                raise Exception("Provider not supported")

    def get_storage_service(self) -> CloudStorageService:
        match self.cloud_provider:
            case CloudProvider.Azure:
                return BlobStorageService(STORAGE_CONNECTION_STRING, AZURE_STORAGE_CONTAINER_NAME)
            case _:
                raise Exception("Provider not supported")
